### PR TITLE
[feature] Combat update

### LINKED ISSUE #719: CrystalAura max bps
Maximum placed crystals per second to prevent wasting crystals

### LINKED ISSUE #772: Explosion Behavior of "VULNERABLE_PLAYERS" for CrystalAura
**Describe the solution you'd like**
The CrystalAura only places and explodes crystals if detects people that are not in holes. Since we have HoleESP thing, it probably could be united to see if person is safe. If we find that it's in a vulnerable place, then CrystalAura starts placing crystals and blowing them up. Simple as that.

### LINKED ISSUE #800: CrystalAura doesn't place first crystal
Title. You need to manually place 1 crystal for it to start placing.

### LINKED ISSUE #836: OffHandGap overrides manual offhand Crystal
**Describe the bug**
Using OffhandGap + AutoTotem > Custom Item > Crystal, wont let you place crystals manually

**To Reproduce**
1. Go to 'menu'
2. Click on 'OffhandGap, enable everything inside OffhandGap Module, even crystal check | set AutoTotem to custom item > Crystals '
3. See error

**Expected behavior**
Well, when you right click, OffhandGap is basically changing your offhand crystal to a gapple, so you cant place the crystal, idk honestly how it could be fixed, i have a toggle with Aura + OffhandGap, so i just disable Aura and enable CA, i shouldnt be doing that, so thats why i think its a bug.

**Additional context**

### LINKED ISSUE #849: Add a MaxSelfDamage option to CrystalAura
Add a MaxDamage option in the crystalaura settings that will not place a crystal if it does more than the set damage to you. For example, if a crystal deals more than 2 hearts on me, I will not place it. MaxDamage will be prioritized over Minimum Damage for enemies. If set to 20, it is essentially off.

### LINKED ISSUE #943: OffhandGap should check for spawned in crystals in range, and then cancel

### LINKED ISSUE #989: Rare offhandgap crash when using crystals
https://pastebin.com/dQWtiVcC

### LINKED ISSUE #1005: AutoTotem Custom Item should support any minecraft item
**Is your feature request related to a problem? Please describe.**
Its not much of a problem but I know a lot of people would like this feature as it covers more scenarios.

**Describe the solution you'd like**
I want autototem, to have full custom item choice. So you can set whatever item you want for autototem. Maybe make this replace either gaps or crystals or add a separate category for it? Just the ability to pick any item you want to be auto offhanded.

**Describe alternatives you've considered**
This seems like the most straightforward solution.

**Additional context**
No additional context.

### LINKED ISSUE #1045: Customization options for OffhandGap
**Is your feature request related to a problem? Please describe.**
The limited use of a utility that could be more useful if given more options

**Describe the solution you'd like**
Inclusion of an option to input custom items for the OffhandGap module.

**Describe alternatives you've considered**
There could possibly be a name change for the OffhandGap module if customizability is included.

**Additional context**
N/A

### LINKED ISSUE #1049: Add a MultiTask module.
Alrighty, so this MultiTask module does a very simple thing.

When you have a Crystal in your  offhand, and any sort of item, both of the actions of those items can be made. For example:

Crystal in Offhand - Pickaxe in Mainhand - You mine enemies, and place crystals on the obsidian you're mining at the same time.

Crystal in Offhand - Gapple in Mainhand - You wanna faceplace someone, but dont have the right CrystalAura for it, well you can eat a gapple and crystal people manually at the same time! (Of course there would be a Explode-only crystalaura enabled, without place enabled)

### LINKED ISSUE #1072: Obsidian miner for crystal pvp
**Is your feature request related to a problem? Please describe.**
No promlem just idea
**Describe the solution you'd like**
I would like possibly just a built in feature to the auto cryztal mechanic in kami blue or a whole new feature as a whole i would like a feature that would mine out obsidian boxed around enemies who cryztal aura is  targetting
**Describe alternatives you've considered**
Possibly somthing that will mineout all types of blocks around people possibly when they are mining away amd using baritone to block up there path.
**Additional context**
Nope

### LINKED ISSUE #1174: CrystalAura + Freecam = oops
**Describe the bug**
Related to #1040, activating Freecam with CrystalAura enabled places crystals on myself.

**To Reproduce**
1. Turn on CrystalAura with "Place" on
2. Turn on Freecam
3. Have a heart attack?

**Expected behavior**
Crystals should not be placed on the freecaming player.

**Screenshots**
Don't think it's necessary.

**Additional context**

### LINKED ISSUE #1212: Bedaura
**Is your feature request related to a problem? Please describe.**
Im always frustrated when i get bed bombed because they hurt alot
**Describe the solution you'd like**
I would like a bed boming feature or bed aura with some sort of auto refilling mechanic where it would replace them whenever they are placed
**Describe alternatives you've considered**
Just a simple bed aura that doesnt re fill the beds but this would mostlikely be much less effective
**Additional context**
https://www.youtube.com/watch?v=x4tAdQ_OLWg&t=76s this is where i got the idea from

### LINKED ISSUE #1262: Aimbot aims at feet
**Describe the bug**

Kami Blue's aimbot module aims at the feet of a player (or more accurately the center of the bottom face of an entity's hitbox) instead of their head or body.

**To Reproduce**

1. turn on aimbot
2. aim at player
3. shoot foot

**Expected behavior**
Aimbot aims at the eye position or the center of the target.

**Screenshots**

image

**Additional context**
At some point in the 1.17 betas the aimbot changed from aiming at the eye position of the target, to the middle of the target, and now to the bottom of the target. I'll figure out the versions in which this change happened sometime tomorrow.

### LINKED ISSUE #1271: AntiBots
Some servers create fake entities / fake players around you to annoy you. Adding a feature to remove them would be cool

### LINKED ISSUE #1329: AutoTotem option to check for nearby crystals against health setting
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

PLZ add offhand crystal and Switch to totems when you take fatal damage, such as jumping out of hole, or when obsidian is dug out from under your feet.

**Describe alternatives you've considered**

**Additional context**

SHANA offhand

### LINKED ISSUE #1377: Aura should attack specific parts of the body and automatically select which part for optimal distance
Two parts, an enhancement and a bug 

First, you should have an option to select a specific part of the body to target (feet, chest, head etc)

Second, Aura currently targets the feet by default afaik, it should attempt to face the optimal target by default.

For example, if you're 1 block higher then someone and 4.3 blocks away, it should attack their head as that's closer then the feet, and so on.

A bit of basic trig can calculate which part of the body is best to target

### PR DESCRIPTION
**Pull request by Xiaro**
Rewrote bunch of combat modules

closes #719 
closes #772 
closes #800 
closes #836 
closes #849 
closes #943
closes #989
closes #1005 
closes #1045 
closes #1049 
closes #1072 
closes #1174
closes #1212 
closes #1262 
closes #1271 
closes #1329 
closes #1377

### CODE REVIEW

**Comment by sourTaste000** [comment]
Does't place crystal until you holds it

**Comment by dominikaaaa** [comment]
> Does't place crystal until you holds it
That's intentional

**Comment by Xiaro** [comment]
> Does't place crystal until you holds it

It is either because that it doesn't find a pos to do enough damage to the target or find a pos that does less enough damage to you

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Surround.kt [hunk 1]
```diff
@@ -165,6 +167,7 @@ object Surround : Module() {
         return if (autoCenter.value == AutoCenterMode.OFF) {
             true
         } else {
+            if (disableStrafe.value) Strafe.disable()
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/combat/Surround.kt:170]: shouldn't this reenable strafe on disable?

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/HoleSnap.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        mc.timer.tickLength = 50f
+    }
+
+    override fun onUpdate() {
+        if (SurroundUtils.checkHole(mc.player) != SurroundUtils.HoleType.NONE) {
+            disable()
+            return
+        }
+        findHole()?.toVec3d()?.let {
+            if (disableStrafe.value) Strafe.disable()
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/HoleSnap.kt:36]: see above

Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/movement/HoleSnap.kt:36]: It disables after you got into a hole. There is no reason to enabled strafe in hole

File: src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java [hunk 3]
```diff
@@ -33,6 +40,24 @@
...
     @Shadow public SoundHandler soundHandler;
+    @Shadow public RayTraceResult objectMouseOver;
+    @Shadow public PlayerControllerMP playerController;
+    @Shadow public EntityRenderer entityRenderer;
+
+    @Inject(method = "rightClickMouse", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/entity/EntityPlayerSP;getHeldItem(Lnet/minecraft/util/EnumHand;)Lnet/minecraft/item/ItemStack;"), cancellable = true)
+    public void processRightClickBlock(CallbackInfo ci) {
+        if (CrystalAura.INSTANCE.isActive()) {
+            ci.cancel();
+            for (EnumHand enumhand : EnumHand.values()) {
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: can't you just use `EnumHand.forEach`?

Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: Minecraft code moment, and it is just some code style lol

Comment (dominikaaaa) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: yea, it technically compiles to a forEach anyways, it's just cleaner to use forEach imo

Comment (dominikaaaa) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: yea, it technically compiles to a forEach anyways, it's just cleaner to use forEach imo

Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: Java Array doesn't have `forEach` lol. And enhanced for loop is already clean enough imo

Comment (dominikaaaa) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: cringe java array

Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/mixin/client/MixinMinecraft.java:51]: ♿

**Comment by willngy** [comment]
CrystalAuraRewrite does not work with the AntiHunger module.

**Comment by l1ving** [comment]
lol i merged master without testing if it worked <kannagun:761771166339301397:>

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Surround.kt [hunk 4]
```diff
@@ -1,280 +1,231 @@
...
+                val placingInfo = getPlaceInfo(placed) ?: break
+                placeCount++
+                placed.add(placingInfo.second.offset(placingInfo.first))
+                BlockUtils.doPlace(placingInfo.second, placingInfo.first, placeSpeed.value)
+                if (placeCount >= 4) break
+            }
+            Thread.sleep(100L)
+            placeCount = 0
+            placed.clear()
+        }*/
```
Comment (l1ving) [inline src/main/java/me/zeroeightsix/kami/module/modules/combat/Surround.kt:208]: delete this please

**Review by l1ving — CHANGES_REQUESTED** [review]

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import kotlin.math.max
+
+// TODO: Merge Auto totem into this Y
+// TODO: Merge OffhandGap into this Y
+// TODO: #1329 Y
+// TODO: #1045 Y
+// TODO: #1049 Y
+// TODO: #1005 Y
+// TODO: #943 Y
+// TODO: #836 Y
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:31]: Yes

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AimBot.kt [hunk 6]
```diff
@@ -1,49 +1,29 @@
...
         description = "Automatically aims at entities for you.",
-        category = Module.Category.COMBAT
+        category = Module.Category.COMBAT,
+        modulePriority = 20
 )
 object AimBot : Module() {
-    private val priority = register(Settings.e<EntityPriority>("Priority", EntityPriority.DISTANCE))
-    private val range = register(Settings.floatBuilder("Range").withValue(16.0f).withMinimum(4.0f).withMaximum(24.0f).build())
-    private val useBow = register(Settings.booleanBuilder("UseBow").withValue(true).build())
-    private val ignoreWalls = register(Settings.booleanBuilder("IgnoreWalls").withValue(true).build())
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AimBot.kt:20]: They all in combat setting now

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.kt [hunk 7]
```diff
@@ -26,99 +20,59 @@ import net.minecraft.util.EnumHand
...
-    private val lockView = register(Settings.booleanBuilder("LockView").withValue(false).withVisibility { !multi.value }.build())
-    private val waitTick = register(Settings.floatBuilder("SpamDelay").withMinimum(0.1f).withValue(2.0f).withMaximum(40.0f).withVisibility { delayMode.value == WaitMode.SPAM }.build())
-    private val eat = register(Settings.b("WhileEating", true))
-    private val players = register(Settings.b("Players", true))
-    private val friends = register(Settings.booleanBuilder("Friends").withValue(false).withVisibility { players.value }.build())
-    private val sleeping = register(Settings.booleanBuilder("Sleeping").withValue(false).withVisibility { players.value }.build())
-    private val mobs = register(Settings.b("Mobs", false))
-    private val passive = register(Settings.booleanBuilder("PassiveMobs").withValue(false).withVisibility { mobs.value }.build())
-    private val neutral = register(Settings.booleanBuilder("NeutralMobs").withValue(false).withVisibility { mobs.value }.build())
-    private val hostile = register(Settings.booleanBuilder("HostileMobs").withValue(false).withVisibility { mobs.value }.build())
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.kt:41]: They all in combat setting now

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.kt [hunk 8]
```diff
@@ -26,99 +20,59 @@ import net.minecraft.util.EnumHand
...
-    private val invisible = register(Settings.b("Invisible", false))
-    private val ignoreWalls = register(Settings.b("IgnoreWalls", true))
-    private val minExistTime = register(Settings.integerBuilder("MinExistTime(s)").withValue(5).withRange(0, 20).build())
-    private val range = register(Settings.f("Range", 5.5f))
-    private val sync = register(Settings.b("TPSSync", false))
-    private val pauseBaritone: Setting<Boolean> = register(Settings.b("PauseBaritone", true))
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/Aura.kt:47]: They all in combat setting now

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AntiFriendHit.kt [hunk 9]
```diff
@@ -16,10 +16,7 @@ import net.minecraft.client.entity.EntityOtherPlayerMP
 object AntiFriendHit : Module() {
     @EventHandler
     private val listener = Listener(EventHook { event: ClientPlayerAttackEvent ->
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AntiFriendHit.kt:18]: No?

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        description = "Manages item in your offhand",
+        category = Module.Category.COMBAT
+)
+object AutoOffhand : Module() {
+    private val type = register(Settings.enumBuilder(Type::class.java, "Type").withValue(Type.TOTEM))
+
+    // Totem
+    private val hpThreshold = register(Settings.floatBuilder("HpThreshold").withValue(5f).withRange(1f, 20f).withVisibility { type.value == Type.TOTEM })
+    private val checkDamage = register(Settings.booleanBuilder("CheckDamage").withValue(true).withVisibility { type.value == Type.TOTEM })
+    private val mob = register(Settings.booleanBuilder("Mob").withValue(true).withVisibility { type.value == Type.TOTEM && checkDamage.value })
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:43]: It is right underneath the check damage

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/CrystalAura.kt [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+        PlayerPacketManager.resetHotbar()
+    }
+
+    @EventHandler
+    private val receiveListener = Listener(EventHook { event: PacketEvent.Receive ->
+        if (!CombatManager.isOnTopPriority(this) || mc.player == null || event.packet !is SPacketSoundEffect) return@EventHook
+        if (event.packet.getCategory() == SoundCategory.BLOCKS && event.packet.getSound() == SoundEvents.ENTITY_GENERIC_EXPLODE) {
+            val crystalList = CrystalUtils.getCrystalList(Vec3d(event.packet.x, event.packet.y, event.packet.z), 5f)
+            for (crystal in crystalList) {
+                crystal.setDead()
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/CrystalAura.kt:134]: No, it is just a weird way of making the CA faster

[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/CrystalAura.kt:134]: Leave a comment in the code explaining why it's there then

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:35]: I still can't tell what it does

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AntiFriendHit.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AntiFriendHit.kt:18]: just wanted to be sure it works

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:35]: Whether checks damage from mob or not. Isn't it clear?

[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:35]: Call it CheckMob or something? :kannagun:

[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:35]: It only displays when check damage is one and I want to keep the setting names short :kannagun:
Sub settings would makes things much easier :woeis:

[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/module/modules/combat/AutoOffhand.kt:35]: Okay I'll fix it when we add subsettings.

**Review by l1ving — APPROVED** [review]

**Pull request merged**